Added Notebooks Server tab in Projects page

lorenzo-cavazzi opened the pull request:
First steps of #351

Not sure if I should move the `fetchNotebookServers` logic inside the `NotebookServers` component. I also have some doubts about a few choices I have made, I'll add some comment in the code

ableuler commented on src/notebooks/Notebooks.container.js:
```diff
@@ -165,4 +170,23 @@ class LaunchNotebookServer extends Component {
+
+  async doStopServer(serverName) {
+    const url = `${this.props.url}/services/notebooks/servers/${serverName}`;
+    await fetch(url, {credentials: 'include', method: 'DELETE'});
```
It would be nice to do the actual fetch from the `api-client` module and have a `stopServer(serverName)` (or similar) method there.

ableuler commented on src/api-client/notebook-servers.js:
```diff
@@ -0,0 +1,36 @@
+    let headers = client.getBasicHeaders();
+    headers.append('Content-Type', 'application/json');
+
+    return client.clientFetch(`${client.jupyterhubUrl}/services/notebooks/user`, {
```
This call should be routed through the gateway which will exchange the session cookie for an access token that jupyterhub accepts. `/api/notebooks/...` will be mapped to `/jupyterhub/services/notebooks/...` there, so it should be enough to adapt the URL in the above call accordingly.

ableuler commented on src/project/Project.state.js:
```diff
@@ -64,6 +64,15 @@ class ProjectModel extends StateModel {
+    this.set('core.notebookServersUpdating', true);
+    return client.getNotebookServers()
+      .then(resp => {
+        this.set('core.notebookServers', resp.data);
```
We might want to revisit the schema for the internal representation of a project soon, but for the moment i think it's ok to store the information like this.

ableuler commented:
Nice job on your first issue! I already have a few requests for changes and I will take another look at it later today.

lorenzo-cavazzi commented:
@ableuler can't answer to your previous comment (now marked as outdated) but I have addressed both

ciyer commented:
This looks very good. The only thing is, it seems that the list over servers is not updated after the stop notebook server command returns. Could this be fixed?

lorenzo-cavazzi commented:
@ciyer Yes, I can poll to check the new status. At the moment we don't have a websocket-like API. Do you think it makes sense to continuously check the list of notebooks while on that page - let's say query the API every 2-5 seconds?

ciyer commented:
I think this makes sense when the user has triggered a "Stop Notebook" -- in this case, we know that there should be a change to the status coming. Once the status has been updated, we can also stop the polling.

lorenzo-cavazzi commented:
Apparently solving that problem is not as easy as I hoped... would it make sense to move it to another issue?

ciyer commented:
The problem is that the "Connect to Jupyter" button is still enabled after the server has stopped?

ciyer commented on src/project/Project.state.js:
```diff
@@ -64,6 +64,15 @@ class ProjectModel extends StateModel {
   }
 
+  fetchNotebookServers(client) {
+    this.set('core.notebookServersUpdating', true);
```
I think setUpdating is better -- it more clearly communicates the intention and is something that can easily be searched for, etc.

lorenzo-cavazzi replied:
changed it so that the first time I query the list of server I use `this.setUpdating`, while the next times I don't want any visual loading effect - and of course I got rid of  `core.notebookServersUpdating`.

lorenzo-cavazzi commented:
> The problem is that the "Connect to Jupyter" button is still enabled after the server has stopped?

yes, and the logic to disable the button is in the `componentDidUpdate` method of the `LaunchNotebookButton` react component.
I have tried some quick workaround but without success. Should that logic be moved somewhere else so that the server status computation can be invoked from outside the component?

rokroskar commented:
There are several components at play here -- is the `Connect to Jupyter` button visible because Jupyterhub (resp. the notebooks service) says that it is running?

rokroskar commented:
> would it make sense to move it to another issue?

IMO yes because this might be a deeper problem involving the notebooks service/jupyterhub.

lorenzo-cavazzi commented:
> There are several components at play here -- is the `Connect to Jupyter` button visible because Jupyterhub (resp. the notebooks service) says that it is running? Or is it really a UI issue (some component not getting updated correctly)?

This is primarily a UI issue because the component is not updated correctly when the server is not running anymore, but if I invoke the API I see the server is not running anymore. On the other hand we don't have a websocket signaling for server status changes, at the moment we can only invoke the API every x seconds to check the real current status - maybe this can be improved.

lorenzo-cavazzi commented:
> > would it make sense to move it to another issue?
> 
> IMO yes because this might be a deeper problem involving the notebooks service/jupyterhub.

If I query the API to check the server status when the server has already stopped, it spawns again. This is a bit strange... I can check the status with another API but it's a bit more complicated. Not sure if it's a problem with jupyterhub

ciyer commented on src/notebooks/Notebooks.present.js:
```diff
@@ -117,4 +119,53 @@ class RangeOption extends Component {
+        {loader}
+      </td>
+      <td>
+        <a className={className} role="button" href={url}>Connect</a>
```
The "Connect" button should open JupyterLab in a new window/tab. Use ```utils/UIComponents.js:ExternalLink``` to get this behavior.

lorenzo-cavazzi replied:
Using `<ExternalLink>` now. I had to slightly change the component so that it can be easily disabled

rokroskar commented:
> If I query the API to check the server status when the server has already stopped, it spawns again.

the spawn should only happen on a `POST` -- if it happens on a `GET` it's not intended and should be fixed.

lorenzo-cavazzi commented:
> > If I query the API to check the server status when the server has already stopped, it spawns again.
> 
> the spawn should only happen on a `POST` -- if it happens on a `GET` it's not intended and should be fixed.

I was using `POST` instead of `GET` also to query the status, which according to [the documentation](https://github.com/SwissDataScienceCenter/renku-notebooks#endpoints) is wrong. But using `GET` is not returning a response in `JSON` format, instead I get an HTML page with data available as a JS object I can easily identify by the id `jupyter-config-data` (see below for an example).
I can live with that, just want to be sure it's the intended behavior.

```
<!doctype html><html lang="en"><head><meta charset="utf-8"><title>JupyterLab</title><meta name="viewport" content="width=device-width, initial-scale=1"><script id="jupyter-config-data" type="application/json">{
    "buildAvailable": "True",
    "buildCheck": "True",
    "token": "",
    "devMode": "False",
    "bundleUrl": "/jupyterhub/user/lorenzo.cavazzi.tech/lorenzo-2e-zurich-2db-24d39f9/lab/static/",
    "notebookVersion": "[5, 6, 0]",
    "terminalsAvailable": "True",
    "ignorePlugins": "[]",
    "serverRoot": "~/zurich-bikes-tutorial",
    "mathjaxConfig": "TeX-AMS-MML_HTMLorMML-full,Safe",
    "mathjaxUrl": "/jupyterhub/user/lorenzo.cavazzi.tech/lorenzo-2e-zurich-2db-24d39f9/static/components/MathJax/MathJax.js",
    "appName": "JupyterLab",
    "appNamespace": "jupyterlab",
    "appSettingsDir": "/opt/conda/share/jupyter/lab/settings",
    "appVersion": "0.35.4",
    "cacheFiles": "True",
    "pageUrl": "/lab",
    "publicUrl": "/lab/static/",
    "schemasDir": "/opt/conda/share/jupyter/lab/schemas",
    "settingsUrl": "/lab/api/settings/",
    "staticDir": "/opt/conda/share/jupyter/lab/static",
    "templatesDir": "/opt/conda/share/jupyter/lab/static",
    "themesDir": "/opt/conda/share/jupyter/lab/themes",
    "themesUrl": "/lab/api/themes/",
    "treeUrl": "/lab/tree/",
    "userSettingsDir": "/home/jovyan/.jupyter/lab/user-settings",
    "workspacesApiUrl": "/lab/api/workspaces/",
    "workspacesDir": "/home/jovyan/.jupyter/lab/workspaces",
    "workspacesUrl": "/lab/workspaces/",
    "baseUrl": "/jupyterhub/user/lorenzo.cavazzi.tech/lorenzo-2e-zurich-2db-24d39f9/",
    "wsUrl": ""
  }</script><link rel="icon" type="image/x-icon" href="/jupyterhub/user/lorenzo.cavazzi.tech/lorenzo-2e-zurich-2db-24d39f9/static/base/images/favicon.ico" class="idle favicon"><link rel="" type="image/x-icon" href="/jupyterhub/user/lorenzo.cavazzi.tech/lorenzo-2e-zurich-2db-24d39f9/static/base/images/favicon-busy-1.ico" class="busy favicon">
  

</head><body><script type="text/javascript">
  /* Remove token from URL. */
  (function () {
    var location = window.location;
    var search = location.search;

    // If there is no query string, bail.
    if (search.length <= 1) {
      return;
    }

    // Rebuild the query string without the `token`.
    var query = '?' + search.slice(1).split('&')
      .filter(function (param) { return param.split('=')[0] !== 'token'; })
      .join('&');

    // Rebuild the URL with the new query string.
    var url = location.origin + location.pathname +
      (query !== '?' ? query : '') + location.hash;

    window.history.replaceState({ }, '', url);
  })();
</script><script type="text/javascript" src="/jupyterhub/user/lorenzo.cavazzi.tech/lorenzo-2e-zurich-2db-24d39f9/lab/static/vendors~main.57bd3cc0bd3cf189076d.js"></script><script type="text/javascript" src="/jupyterhub/user/lorenzo.cavazzi.tech/lorenzo-2e-zurich-2db-24d39f9/lab/static/main.6bdd2f1e3af6b3a3a469.js"></script></body></html>
```

rokroskar commented:
You need to ask for `application/json` in the header

lorenzo-cavazzi commented:
I have already added it
```
x-requested-with:XMLHttpRequest
Content-Type:application/json
Cookie:<mycookie>
```

rokroskar commented:
And you get html instead of json?

lorenzo-cavazzi commented:
yes

rokroskar commented:
@lorenzo-cavazzi apologies I didn't read carefully earlier -- your header should be `Accept: application/json` not `Content-Type: application/json`.

lorenzo-cavazzi commented:
I have expanded #340 so that the scope of this PR won't become way bigger than the original issue it references. The "Connect to Jupiter" button problem will be address there ~~*immediately*~~ soon

lorenzo-cavazzi commented:
closes #351
